# HG changeset patch
# User Michael Goffioul <michael.goffioul@gmail.com>
# Date 1242936728 -3600
#      Thu May 21 21:12:08 2009 +0100
# Node ID c02224afead667683be30e023f0212ceaa457b85
# Parent  b0395373253052d1795f9cfae9fae86341ef9a30
Add mkstemp implementation based on mktemp, for platforms missing it

diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,10 +1,14 @@
 2009-05-21  Michael Goffioul  <michael.goffioul@gmail.com>
 
+	* file-io.cc (mkstemp): Add mktemp-based implementation of mkstemp on
+	platforms that do not have it (mkstemp is required by the new help
+	system).
+
 	* TEMPLATE-INST/Array-os.cc: Add "extern template" declaration for
 	Array<octave_idx_type> to avoid implicit instantiation (and duplicate
 	symbols at link-time) [Win32]
 
 2009-05-20  John W. Eaton  <jwe@octave.org>
 
 	* pt-assign.cc (maybe_warn_former_built_in_variable):
 	Set initialized to true when done initializing vars set.
diff --git a/src/file-io.cc b/src/file-io.cc
--- a/src/file-io.cc
+++ b/src/file-io.cc
@@ -1907,16 +1907,34 @@ system-dependent error message.\n\
   return retval;
 }
 
 #if defined (HAVE_MKSTEMPS)
 // Prototype for mkstemps in libiberty
 extern "C" int mkstemps (char *pattern, int suffix_len);
 #endif
 
+#if ! defined (HAVE_MKSTEMP) && ! defined (HAVE_MKSTEMPS) && defined (_MSC_VER)
+# if defined (HAVE_FCNTL_H)
+#  include <fcntl.h>
+# endif
+# if defined (HAVE_SYS_STAT_H)
+#  include <sys/stat.h>
+# endif
+int mkstemp (char *tmpl)
+{
+  int ret=-1;
+  mktemp (tmpl);
+  ret = open (tmpl, O_RDWR | O_BINARY | O_CREAT | O_EXCL | _O_SHORT_LIVED,
+	      _S_IREAD | _S_IWRITE);
+  return ret;
+}
+#define HAVE_MKSTEMP 1
+#endif
+
 DEFUN (mkstemp, args, ,
   "-*- texinfo -*-\n\
 @deftypefn {Built-in Function} {[@var{fid}, @var{name}, @var{msg}] =} mkstemp (@var{template}, @var{delete})\n\
 Return the file ID corresponding to a new temporary file with a unique\n\
 name created from @var{template}.  The last six characters of @var{template}\n\
 must be @code{XXXXXX} and these are replaced with a string that makes the\n\
 filename unique.  The file is then created with mode read/write and\n\
 permissions that are system dependent (on GNU/Linux systems, the permissions\n\
